File: clients/csharp/Model/Action.cs
namespace SpbAiChamp.Model
{
    /// <summary>
    /// TODO - Document
    /// </summary>
    public struct Action
    {
        /// <summary>
        /// TODO - Document
        /// </summary>
        public Model.MoveAction[] Moves { get; set; }
        /// <summary>
        /// TODO - Document
        /// </summary>
        public Model.BuildingAction[] Buildings { get; set; }
    
        public Action(Model.MoveAction[] moves, Model.BuildingAction[] buildings)
        {
            this.Moves = moves;
            this.Buildings = buildings;
        }
    
        /// <summary> Read Action from reader </summary>
        public static Action ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new Action();
            result.Moves = new Model.MoveAction[reader.ReadInt32()];
            for (int movesIndex = 0; movesIndex < result.Moves.Length; movesIndex++)
            {
                result.Moves[movesIndex] = Model.MoveAction.ReadFrom(reader);
            }
            result.Buildings = new Model.BuildingAction[reader.ReadInt32()];
            for (int buildingsIndex = 0; buildingsIndex < result.Buildings.Length; buildingsIndex++)
            {
                result.Buildings[buildingsIndex] = Model.BuildingAction.ReadFrom(reader);
            }
            return result;
        }
    
        /// <summary> Write Action to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(Moves.Length);
            foreach (var movesElement in Moves)
            {
                movesElement.WriteTo(writer);
            }
            writer.Write(Buildings.Length);
            foreach (var buildingsElement in Buildings)
            {
                buildingsElement.WriteTo(writer);
            }
        }
    
        /// <summary> Get string representation of Action </summary>
        public override string ToString() {
            string stringResult = "Action { ";
            stringResult += "Moves: ";
            stringResult += "[ ";
            int movesIndex = 0;
            foreach (var movesElement in Moves)
            {
                if (movesIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += movesElement.ToString();
                movesIndex++;
            }
            stringResult += " ]";
            stringResult += ", ";
            stringResult += "Buildings: ";
            stringResult += "[ ";
            int buildingsIndex = 0;
            foreach (var buildingsElement in Buildings)
            {
                if (buildingsIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += buildingsElement.ToString();
                buildingsIndex++;
            }
            stringResult += " ]";
            stringResult += " }";
            return stringResult;
        }
    }
}

File: clients/go/model/action.go
package model

import "io"
import . "spb_ai_champ/stream"

// TODO - Document
type Action struct {
    // TODO - Document
    Moves []MoveAction
    // TODO - Document
    Buildings []BuildingAction
}

func NewAction(moves []MoveAction, buildings []BuildingAction) Action {
    return Action {
        Moves: moves,
        Buildings: buildings,
    }
}

// Read Action from reader
func ReadAction(reader io.Reader) Action {
    var moves []MoveAction
    moves = make([]MoveAction, ReadInt32(reader))
    for movesIndex := range moves {
        var movesElement MoveAction
        movesElement = ReadMoveAction(reader)
        moves[movesIndex] = movesElement
    }
    var buildings []BuildingAction
    buildings = make([]BuildingAction, ReadInt32(reader))
    for buildingsIndex := range buildings {
        var buildingsElement BuildingAction
        buildingsElement = ReadBuildingAction(reader)
        buildings[buildingsIndex] = buildingsElement
    }
    return Action {
        Moves: moves,
        Buildings: buildings,
    }
}

// Write Action to writer
func (action Action) Write(writer io.Writer) {
    moves := action.Moves
    WriteInt32(writer, int32(len(moves)))
    for _, movesElement := range moves {
        movesElement.Write(writer)
    }
    buildings := action.Buildings
    WriteInt32(writer, int32(len(buildings)))
    for _, buildingsElement := range buildings {
        buildingsElement.Write(writer)
    }
}

// Get string representation of Action
func (action Action) String() string {
    stringResult := "{ "
    stringResult += "Moves: "
    moves := action.Moves
    stringResult += "[ "
    for movesIndex, movesElement := range moves {
        if movesIndex != 0 {
            stringResult += ", "
        }
        stringResult += movesElement.String()
    }
    stringResult += " ]"
    stringResult += ", "
    stringResult += "Buildings: "
    buildings := action.Buildings
    stringResult += "[ "
    for buildingsIndex, buildingsElement := range buildings {
        if buildingsIndex != 0 {
            stringResult += ", "
        }
        stringResult += buildingsElement.String()
    }
    stringResult += " ]"
    stringResult += " }"
    return stringResult
}